### PR TITLE
Kube resource pinning

### PR DESCRIPTION
**Pull request by grantnicholas**
This PR deals with crash safety of the scheduler. We do this by periodically persisting the resourceVersion of the watch into the airflow metadata database. When the scheduler starts up it then uses that resourceVersion from the metadata database to start watching.

### CODE REVIEW

File: airflow/contrib/executors/kubernetes_executor.py [hunk 1]
```diff
@@ -145,42 +144,56 @@ def __init__(self, namespace, watcher_queue):
...
         self.namespace = namespace
         self.watcher_queue = watcher_queue
-        self._api = client.CoreV1Api()
-        self._watch = watch.Watch()
 
     def run(self):
+        from airflow.models import KubeResourceVersion
+        from airflow import settings
+
+        kube_client = get_kube_client()
```
Comment (dimberman) [inline airflow/contrib/executors/kubernetes_executor.py]: Do you think that we should have some form of start-up script that checks for still-pending tasks that are still pending and kill them? I would imagine that tasks that are still running wouldn't submit their resourceVersion. (this also raises the concerning thought that a task that was pending when the scheduler died could start running during the restart)

File: airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+from alembic import op
+import sqlalchemy as sa
+
+
+RESOURCE_TABLE = "kube_resource_version"
+
+
+def upgrade():
```
Comment (dimberman) [inline airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py]: One concern I'm having is that if the postgres instance is part of the airflow deployment, and that gets torn down too, wouldnt you lose that information anyways? Is there no way to query kubernetes for the most recent resourceVersion with certain tags?

**Comment by dimberman** [comment]
So how should we go about setting up the testing for this? I imagine we should have something in the integration testing that can spin-up/tear down the airflow instances multiple times

File: airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py (hunk 2 above)
Comment (grantnicholas) [reply airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py]: The deployment going away is fine because the data will still exist when the database deployment is recreated (as long as people don't delete the backing persistent volume which they shouldn't).

About the second bit. We actually don't want the most recent resourceVersion for our label selector. We want the last resourceVersion that we saw and processed. The distinction being that for every resourceVersion event from kubernetes our executor needs to perform an action. Only *after* we have performed that action for that event do we bump the resourceVersion in the database. This is a way to provide (almost) "exactly once" processing of events even when the executor is temporarily down.

File: airflow/contrib/executors/kubernetes_executor.py (hunk 1 above)
Comment (grantnicholas) [reply airflow/contrib/executors/kubernetes_executor.py]: Good question. As we can guarantee we are reading and processing every event from kubernetes once, there should be no issues with the scheduler being offline for a very long time. We know we can always "start from where we left off" in the event stream. 

That said, it gets more complicated when you start wanting to do client timeout based health checking (like discussed in slack). In that case you're right. Saving a single resourceVersion is not enough and you need to do extra work to maintain the invariant that *I receive events shortly after they happen* so you can use a client side timer to approximate a server side timer.

**Comment by grantnicholas** [comment]
You are right that testing this is hard. There are two main approaches:

1. Randomly delete the airflow pod in the middle of dag execution (what you said). We should definitely do this, the downside being we might miss some edge cases. 
2. Try to purposefully trigger bad edge cases with targeted tests. For example: forcibly put a pod in a pending state forever and then delete the airflow pod and see what happens when it recovers. These are more time consuming to write but will provide more confidence about specific behaviors. 

Additionally, do we want these tests integrated into the airflow CI process? If so that will take additional work. Do you know who I should talk with in the airflow community about their CI system and test running?

File: airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py (hunk 2 above)
Comment (dimberman) [reply airflow/migrations/versions/33ae817a1ff4_add_kubernetes_resource_checkpointing.py]: Oh perfect. Another option might be making the DB a stateful set, but as long as that's been covered we should be good.

**Comment by grantnicholas** [comment]
One other thing:
There is a bug (or technically feature omission) with airflow right now that makes all schedulers not crash safe. Tasks get put into the `queued` state by the scheduler and then handed off to the executor. The executor runs `execute_async` and launches the task and then until the task instance is started and the state is set to `running` the task state is still in `queued`. 

This leads to a problem when crashes of the scheduler occur because if a task is `queued` we don't know if the task  has been launched by the executor or not.  

Airflow knows about this because I see this in `state.py`:
```
    # set by the executor (t.b.d.)
    # LAUNCHED = "launched"
``` 

Technically if you want to be crash safe you should set the state from `queued` -> `launched` when `execute_async` is called. Then on startup of the scheduler, we know that tasks that are `queued` should be executed while tasks that are `launched` should not be.  

Since this `launched` state does not exist, I did the hack of resetting the state of all `queued` tasks to `None` on scheduler startup (which will cause the scheduler to reschedule those tasks). Since we don't know whether the `queued` tasks were actually launched or not, this is the only thing we can do. It means that there could be duplicate launched tasks. However, there are some guards in the `TaskInstance.run` method that prevent running of duplicate tasks, so overall even though we might launch two duplicate tasks only one will succeed.

File: airflow/contrib/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -343,33 +343,47 @@ def __init__(self):
         self.task_queue = None
         self._session = None
         self.result_queue = None
-        self.kub_client = None
+        self.kube_scheduler = None
+        self.kube_client = None
         super(KubernetesExecutor, self).__init__(parallelism=self.kube_config.parallelism)
 
-    def clear_queued(self):
+    def clear_not_launched_queued_tasks(self):
```
Comment (grantnicholas) [inline airflow/contrib/executors/kubernetes_executor.py]: @dimberman made the changes here we talked about (only clearing not-launched queued tasks on restart of the scheduler)

**Comment by grantnicholas** [comment]
@dimberman Just bumping this to see what other changes we want to make. Likewise, should I share this with Ani (@foxish?) or others?

**Comment by dimberman** [comment]
@grantnicholas Let me review now. @foxish is gonna be pretty slammed this week with the k8s 1.8 release, but I think having him review next week (if he has time) is a good idea.

File: airflow/contrib/executors/kubernetes_executor.py (hunk 3 above)
Comment (dimberman) [reply airflow/contrib/executors/kubernetes_executor.py]: @grantnicholas So just to be certain, if we put it back in State.NONE airflow will assume that it's ready to be queued and start the whole process over?

Comment (grantnicholas) [reply airflow/contrib/executors/kubernetes_executor.py]: Yup, exactly. This is tested by the integration test that kills the scheduler randomly in the middle of execution and ensures the job succeeds after airflow restarts.

One other bit is: we could technically perform these kubernetes client calls using the async flag to improve performance (ie: perform all the rest API calls in parallel). I just didn't bother to do it since this method is only called once on startup.
